File: util/anim/EAnimD.java
package common.util.anim;

import common.CommonStatic;
import common.system.P;
import common.system.fake.FakeGraphics;

@SuppressWarnings("ForLoopReplaceableByForEach")
public class EAnimD<T extends Enum<T> & AnimI.AnimType<?, T>> extends EAnimI {

	public T type;

	protected MaAnim ma;

	public float f = -1f;

	public EAnimD(AnimI<?, T> ia, MaModel mm, MaAnim anim, T t) {
		super(ia, mm);
		ma = anim;
		type = t;
	}

	@SuppressWarnings("unchecked")
	public void changeAnim(T t, boolean skip) {
		f = -1;
		ma = ((AnimD<?, T>) anim()).getMaAnim(t);
		type = t;
		if (skip)
			setTime(0);
	}

	public boolean done() {
		return f >= ma.max;
	}

	@Override
	public void draw(FakeGraphics g, P ori, float siz) {
		if (f == -1) {
			f = 0;
			setup();
		}
		set(g);
		g.translate(ori.x, ori.y);
		for (EPart e : order) {
			P p = P.newP(siz, siz);
			e.drawPart(g, p);
			P.delete(p);
		}
	}

	/**
	 * Draw parts with specific opacity and size
	 * @param g Graphic
	 * @param ori Position
	 * @param siz Size
	 * @param opacity Opacity, range is 0 ~ 255
	 */
	public void drawBGEffect(FakeGraphics g, P ori, float siz, int opacity, float sizX, float sizY) {
		if(f == -1) {
			f = 0;
			setup();
		}
		set(g);
		g.translate(ori.x, ori.y);
		if(CommonStatic.getConfig().ref) {
			g.colRect(0, 0, 2, 2, 0, 255, 0, 255);
		}
		for (int i = 0; i < order.length; i++) {
			P p = P.newP(siz * sizX, siz * sizY);
			order[i].drawPartWithOpacity(g, p, opacity);
			P.delete(p);
		}
	}

	public float getBaseSizeX() {
		return mamodel.parts[0][8] * 1f / mamodel.ints[0];
	}

	public float getBaseSizeY() {
		return mamodel.parts[0][9] * 1f / mamodel.ints[0];
	}

	public void removeBasePivot() {
		mamodel.parts[0][6] = 0;
		mamodel.parts[0][7] = 0;
	}

	@Override
	public float ind() {
		return f;
	}

	@Override
	public int len() {
		return ma.max + 1;
	}

	@Override
	public void setTime(float value) {
		setup();
		f = value;
		ma.update(f, this, true);
	}

	public void setup() {
		ma.update(0, this, false);
	}

	@Override
	public void update(boolean rotate) {
		if (CommonStatic.getConfig().performanceModeAnimation) {
			if (f == -1) {
				f++;
			} else {
				f += 0.5f;
			}
		} else {
			f++;
		}

		ma.update(f, this, rotate);
	}

	@Override
	protected void performDeepCopy() {
		super.performDeepCopy();
		((EAnimD<?>) copy).setTime(f);
	}

	/**
	 * make this animation a component of another, used in warp and kb
	 */
	public void paraTo(EAnimD<?> base) {
		if (base == null)
			ent[0].setPara(null);
		else
			ent[0].setPara(base.ent[1]);
	}
}

File: util/anim/EAnimD.kt
package common.util.anim

import common.battle.data.DataEntity
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.system.P
import common.system.fake.FakeGraphics
import common.util.ImgCore
import common.util.anim.AnimI.AnimType
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter

open class EAnimD<T>(ia: AnimI<*, T>, mm: MaModel, anim: MaAnim?) : EAnimI(ia, mm) where T : Enum<T>?, T : AnimType<*, T>? {
    var type: T? = null
    protected var ma: MaAnim?
    protected var f = -1
    fun changeAnim(t: T) {
        f = -1
        ma = (anim() as AnimD<*, T>).getMaAnim(t)
        type = t
    }

    fun done(): Boolean {
        return f > ma!!.max
    }

    override fun draw(g: FakeGraphics, ori: P, siz: Double) {
        if (f == -1) {
            f = 0
            setup()
        }
        ImgCore.Companion.set(g)
        g.translate(ori.x, ori.y)
        for (e in order) {
            val p: P = P.Companion.newP(siz, siz)
            e.drawPart(g, p)
            P.Companion.delete(p)
        }
    }

    override fun ind(): Int {
        return f
    }

    override fun len(): Int {
        return ma!!.max + 1
    }

    override fun setTime(value: Int) {
        setup()
        f = value
        ma!!.update(f, this, true)
    }

    fun setup() {
        ma!!.update(0, this, false)
    }

    override fun update(rotate: Boolean) {
        f++
        ma!!.update(f, this, rotate)
    }

    protected override fun performDeepCopy() {
        super.performDeepCopy()
        (copy as EAnimD<*>).setTime(f)
    }

    init {
        ma = anim
    }
}
